include/envoy/upstream/locality_endpoint.h

```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include "envoy/config/core/v3/base.pb.h"
+#include "envoy/config/endpoint/v3/endpoint_components.pb.h"
+
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+using LocalityEndpointTuple = std::tuple<const envoy::config::core::v3::Locality&,
+                                         const envoy::config::endpoint::v3::LbEndpoint&>;
+struct LocalityEndpointHash {
+  size_t operator()(const LocalityEndpointTuple& values) const {
+    const envoy::config::core::v3::Locality& locality = std::get<0>(values);
+    const envoy::config::endpoint::v3::LbEndpoint& endpoint = std::get<1>(values);
+
+    std::string text;
+    {
+      // For memory safety, the StringOutputStream needs to be destroyed before
```

Comment:
Can you move this to `source/common/protobuf/utility.h`? We already have precedence for this kind of utility for `RepeatedPtrField` there. What I would suggest is doing one for `MessageUtil`, hashing each message independently and then hashing together their result.

Reply:
I've simplified this to just hash twice with MessageUtil and XOR them together.